content/blog/2022-05-22-object/index.md

```diff
@@ -0,0 +1,162 @@
+Object-Oriented Programming is a popular programming model that organizes program around objects. It offers some notable expressibility, reusablity, and extensiblity, while not being super complicated. A lot of modern programming languages do support this paradigm. Although Bril is a low-level IR, providing some basic supports to object would make Bril way more expressive and interesting for people to play with, maybe as a tool for potential transpiler that converts a program in high-level language to Bril. This motivates us to add object representation in Bril. A central challenge of this project is that Bril is very limited in terms of its core language syntax, and we have very limited tools to build object upon Bril. We also want to remind ourselves that, Bril is, foundamentally low-level and agnostic to any higher-level language. Therefore, we don't want to add too much complexity to Bril to the extend that new abstractions would become a burden. We have decided to make this extension minimally evasive so that it does not interfere with any Bril's core syntax.
+
+## What I've actually done?
+In this project, I have designed a couple of new primitives, which provide higher-level abstractions. The whole implementation, is basically just a compiler pass which translates a Bril program with object-related syntax to a Bril program with only minimal/core syntax, so that it's completely modular and does not change the behavior of current Bril programs.
+
+## A class-based implementation
+A key question in the design is that whether to use a class-based object model or a prototype-based object model. In a class-based model, objects are pre-defined for their behaviors and types, meanwhile in a prototype-based model, objects are the primary entities, and objects pass all methods and values with each other through delegations. While a prototyped-based model is considered much simpler, I picked the class-based model since Bril is statically-typed and I don't want to add something that break the well-formedness of Bril. It also suits the current Bril toolings more and would get us rid of hectic dynamic typecheckings.
+
+## An example
+Let's start from an extended Bril program, which I would use to explain any new syntax and translation.
+
+``` 
+%rectangle {
+  length: int;
+  width: int:
+
+  @init(input_l: int, input_w: int){
+    lengthref: ref = dot this length;
+    widthref: ref = dot this width;
+
+    set lengthref input_l;
+    set widthref input_w;
+  }
+
+  @area: int{
+    lengthref: ref = dot this length;
+    widthref: ref = dot this width;
+
+    length: int = get lengthref;
+    width: int = get widthref;
+
+    prod: int = mul length width; 
+    ret prod;
+  }
+}
+
+@main {
+  a: int = const 4;
+  b: int = const 2;
+
+  rect1: rectangle = new rectangle;
+  initref: ref = dot rect1 init;
+  call @initref a b;
+  
+  arearef: ref = dot rect1 area;
+  area: int = call @arearef;
+  
+  print area;
+
+  delete rect1;
+}
+```
```

Comment:
Overall this is an interesting design! You have a new type, `ref`, and instructions to manipulate them (`dot` and `get`). This seems like a good approach, overall, to encoding the "soul" of an OO system.

One high-level question: any particular reason why `ref` is not a parameterized type, like `ptr`? It could be `ref<int>`, for example.